api/query_engine.go

```diff
@@ -17,142 +19,326 @@ var (
-	// (starting with raw, then rollups in decreasing precision)
-	// that retains all the data we need and does not exceed minIntervalSoft.
-	// fallback to lowest res option (which *should* have the longest TTL)
-	var found bool
 	for i := range reqs {
 		req := &reqs[i]
-		retentions := mdata.Schemas.Get(req.SchemaId).Retentions.Rets
-		for i, ret := range retentions {
+		var ok bool
+		rets := getRetentions(*req)
+		for i, ret := range rets {
 			// skip non-ready option.
 			if ret.Ready > from {
 				continue
 			}
-			found = true
-			req.Archive = uint8(i)
-			req.TTL = uint32(ret.MaxRetention())
-			if i == 0 {
-				// The first retention is raw data, so use its native interval
-				req.ArchInterval = req.RawInterval
-			} else {
-				req.ArchInterval = uint32(ret.SecondsPerPoint)
-			}
+			ok = true
+			req.Plan(i, ret)
 
-			if req.TTL >= minTTL && req.ArchInterval >= minIntervalSoft {
+			if req.TTL >= minTTL {
 				break
 			}
 		}
-		if !found {
-			return nil, 0, 0, errUnSatisfiable
+		if !ok {
+			return nil, ok
 		}
-
 		if _, ok := seenIntervals[req.ArchInterval]; !ok {
 			listIntervals = append(listIntervals, req.ArchInterval)
 			seenIntervals[req.ArchInterval] = struct{}{}
 		}
 	}
-
-	// due to different retentions coming into play, different requests may end up with different resolutions
-	// we all need to emit them at the same interval, the LCM interval >= interval of the req
 	interval := util.Lcm(listIntervals)
 
-	if interval < minIntervalHard {
-		return nil, 0, 0, errMaxPointsPerReq
+	// plan all our requests so that they result in the common output interval.
+	for i := range reqs {
+		req := &reqs[i]
+		req.AdjustTo(interval, from, getRetentions(*req))
 	}
 
-	// now, for all our requests, set all their properties.  we may have to apply runtime consolidation to get the
-	// correct output interval if out interval != native.  In that case, we also check whether we can fulfill
-	// the request by reading from an archive instead (i.e. whether it has the correct interval.
-	// the TTL of lower resolution archives is always assumed to be at least as long so we don't have to check that)
+	return reqs, true
+}
+
+// note: we can assume all reqs have the same MDP.
+func planLowestResForMDPMulti(now, from, to uint32, reqs []models.Req) ([]models.Req, bool) {
+	var ok bool
+	minTTL := now - from
+
+	// if we were to set each req to their coarsest interval that results in >= MDP/2 points,
+	// we'd still have to align them to their LCM interval, which may push them in to
+	// "too coarse" territory.
+	// instead, we pick the coarsest allowable artificial interval...
+	maxInterval := (2 * (to - from)) / reqs[0].MaxPoints
+	// ...and then we look for the combination of intervals that scores highest.
+	// the bigger the interval the better (load less points), adjusted for number of reqs that
+	// have that interval. but their combined LCM may not exceed maxInterval.
+
+	var validIntervalss [][]uint32
```

Comment:
> I would say that we expect most (if not all) of the series that match a given expression to have the same retentions (since it's pattern based)

yes. most but not all. so we must account for the not case.

>  Is there an optimization available here to just move on if the retentions are identical to a previously evaluated series?

we check if the validIntervals equal any previously seen validIntervals before adding it.
But you mean we could do the check earlier on the entire `rets` rather than the filtered `validIntervals` ?  We could do that. we could check if we've previously processed the same req.SchemaId

Reply:
> yes. most but not all. so we must account for the not case. 

I expect the common case to actually be they are all the same. It definitely is in our setup where we only have a small handful of different retention policies and they are almost never mixed.
I just mean optimize for the common case, not ignore the uncommon case.

Reply:
I also just thought it might be a simplification to separate the concerns. Concern number one would be getting all of the unique retention policies. Concern 2 would then be finding the best match amongst them. Concern 3 would be to set each req appropriately. It's similar to what is already happening, but easier to think about in my head.

Reply:
yep. will do